### ISSUE TITLE
Support for Distinct

### ISSUE THREAD

**Issue by tzachshabtay** [volatiletech/sqlboiler#537]
It would be sweet to have built-in support for the `DISTINCT ON` clause, i.e something like:

```
dbm.Sales(
    qm.Distinct("sales.product_id"),
    qm.Where("sales.product_id = ANY(?)", ids),
    qm.OrderBy("sales.sale_date DESC")
)
```

**Comment by aarondl** [comment]
Would accept a PR for adding a Distinct query mod.

**Comment by mpwalkerdine** [comment]
I tried to workaround this with SELECT + GROUP BY for filtering on associations without selecting them, but ran in to problems with needing to prefix the column names to avoid ambiguity, which then breaks the model binding due to the way `qm.Select` alters construction to include aliases.

e.g.
```go
dbm.Products(
	qm.Select("product.id", "product.name"),
	qm.InnerJoin("sales on product.id = sales.product_id"),
	qm.Where("sales.quantity > ?", 100),
	qm.GroupBy("product.id", "product.name"),
)
```

But the generated query is e.g.

```sql
SELECT "product"."id" as "product.id", "product"."name" as "product.name"
FROM "dbm"."product"
INNER JOIN dbm.sales on product.id = sales.product_id
WHERE (dbm.sales.quantity > $1)
GROUP BY product.id,product.name
```

`qm.Select("id", "name")` can't work in the general case if tables have matching column names (e.g. `id`)

Is there any way to construct this query via the builder API? i.e. either select without the aliases, or saying "only select the `product` table", where it prefixes for you if there are joins e.g.

```sql
SELECT "product"."id", "product"."name" ...
-- or
SELECT "product"."id" as "id", "product"."name" as "name" ...
```

I assume the purpose of the aliases is to catch cases where the actual column names contain e.g. `.`, it's just that specifying a two part name gets caught by the same logic?

**Comment by aarondl** [comment]
The reason select does this is for inner join support because the `Bind` stuff needs to be able to tell which table each column belongs to and if you do: `select "product"."id"` with no alias, the column name returned from the database is just `id` which is of course ambiguous.

Looking at the code it doesn't appear that there's any escape hatch for getting around this if you intend on selecting only a subset of the product's columns. If you wanted all of them I have a feeling `select product.*` would not end up getting modified in that way.

**Comment by mpwalkerdine** [comment]
That's brilliant - I don't intend on selecting a subset of columns, so something like this works:

```
dbm.Products(
	qm.Select("product.*),
	qm.InnerJoin("sales on product.id = sales.product_id"),
	qm.Where("sales.quantity > ?", 100),
        qm.Load("product.sales"),
	qm.GroupBy("product.id", "product.name"),
)
```

For some context, this came up in combination with `Load()`, as only the first of each result line has its `.R` field populated when inner joining. I was looking for a way to query by an associated attribute, then using `Load()` to materialise the resulting object graph, without de-duping the rows client-side.

**Comment by AWhetter** [comment]
Building on the above workaround, it also seems possible to include the DISTINCT in the Select call, so long as you include all the other columns as well.
```
dbm.Products(
	qm.Select("DISTINCT product.id, product.*"),
	qm.InnerJoin("sales on product.id = sales.product_id"),
	qm.Where("sales.quantity > ?", 100),
        qm.Load("product.sales"),
)
```

I'm guessing `Bind` silently ignores unknown columns?

**Comment by aarondl** [comment]
@AWhetter Yes, Bind was made to ignore unknown columns for forward compatibility. If your binary wasn't rebuilt with the latest schema but someone/some other client added a column to a table (a backwards compatible change) then Bind would break.

Thinking about it now it was a change made so we could continue to use `select *` which was probably a mistake, but that's currently how it operates.

**Comment by aarondl** [comment]
Closed via #692

**Comment by sharmaayush358** [comment]
#692 adds support for `DISTINCT`. Support for `DISTINCT ON` would be a great addition. This clause is supported in PostgreSQL. Check the `DISTINCT Clause` section of [this doc](https://www.postgresql.org/docs/current/sql-select.html) for reference.
